hybrid/src/main/java/jetbrains/jetpad/hybrid/ParsingHybridProperty.java

```diff
@@ -66,18 +66,51 @@ public ParsingHybridProperty(
     update();
     mySourceTokens.addListener(new CollectionListener<Token>() {
       @Override
-      public void onItemAdded(CollectionItemEvent<? extends Token> event) {
-        update();
+      public void onItemAdded(final CollectionItemEvent<? extends Token> event) {
+        if (event.getNewItem() instanceof CommentToken) {
+          executeInUpdate(new Runnable() {
+            @Override
+            public void run() {
+              myPrettyTokens.add(event.getIndex(), event.getNewItem());
+            }
+          });
+        } else {
+          update();
+        }
       }
 
       @Override
-      public void onItemSet(CollectionItemEvent<? extends Token> event) {
-        update();
+      public void onItemSet(final CollectionItemEvent<? extends Token> event) {
+        if (event.getNewItem() instanceof CommentToken) {
+          executeInUpdate(new Runnable() {
+            @Override
+            public void run() {
+              myPrettyTokens.set(event.getIndex(), event.getNewItem());
+            }
+          });
+          if (!(event.getOldItem() instanceof CommentToken)) {
+            update();
+          }
+        } else if (event.getOldItem() instanceof CommentToken) {
+          myPrettyTokens.set(event.getIndex(), event.getNewItem());
```

Comment:
Yes, it's missing and should be added here. Tests passed because call returned to `mySourceTokens` set-listener which, apparently, tried to replace new item with itself but this was guarded by `executeInUpdate` inside `update()`. Please review the rest and I'll add corrections later.